### PR TITLE
Fix test_ddp_zero_overlap parity check on 4 or more ranks

### PR DESCRIPTION
**Pull request by albmalamd**
## Summary:

test_ddp_zero_overlap compares a DDP model whose optimizer step is overlapped via ZeroRedundancyOptimizer against a reference DDP model driven by a local SGD, and asserts the parameters match afterwards. The two models were not reducing gradients the same way.

The overlapped model must run 2 warmup iterations (3 with static_graph) because the ZeRO comm hook only starts updating parameters once DDP has rebuilt its gradient buckets, so all of its compared iterations all-reduce over the rebuilt bucket layout. The reference model got no warmup, so its first compared iteration still all-reduced every gradient as a single bucket. For resnet50 on 4 ranks that is one 25557032-element all-reduce against five all-reduces of 2049000/7875584/6563840/6637568/2431040 elements.

RCCL does not sum in the same order for different message layouts: an element's accumulation start position depends on which ring chunk it lands in, which depends on the message size. Floating-point addition is not associative, so the two models get gradients that differ in the low bits. With 2 ranks every ordering collapses to a+b, which is why only 4 or more ranks were affected. Whether the difference appears at all depends on the ranks disagreeing in the first place: the test feeds every rank the same input, and summing identical floats is order-independent. A cold MIOpen cache breaks that, because MIOpen's per-GPU solver selection diverges and different GPUs run different convolution kernels (measured: 46 of 161 gradients not bitwise equal across 4 ranks on a cold cache, 0 of 161 on a warm one). That is the source of the flakiness.

The difference is then amplified out of all proportion because the test trains with loss = output.sum() and lr=0.01, which diverges: conv1.weight grows from 0.4 to 87.6 over the five compared steps. A low-bit gradient difference in the first iteration became an O(10) parameter difference by the last, reported as 71.4% of conv1.weight mismatched.

Give the reference model the same warmup so both models have rebuilt their DDP buckets before the compared iterations and issue an identical sequence of all-reduces. The warmup does not step the optimizer and local_optim.zero_grad() already precedes every compared iteration, so the reference still performs exactly 5 SGD steps on inputs[0..4] from its initial parameters; only its DDP bucket state changes.

## Test Plan:

Clearing the MIOpen kernel cache and perf DB makes the failure deterministic; clearing ~/.cache/miopen alone is not enough. Three cold-cache runs failed before the change with 6719, 6719 and 6720 of 9408 elements mismatched (the ticket reports 6720 of 9408) and passed after it:

```
rm -rf ~/.cache/miopen ~/.config/miopen ~/.cache/comgr
cd test/distributed/optim
python3 -m pytest test_zero_redundancy_optimizer.py -q \
  -k test_ddp_zero_overlap_use_gpu_True_use_interleaved_hook_False_gradient_as_bucket_view_False_static_graph_False_shard_buckets_False
```

All 16 parametrizations, cold cache (16 passed) and warm cache (16 passed):

```
rm -rf ~/.cache/miopen ~/.config/miopen ~/.cache/comgr
cd test/distributed/optim
python3 -m pytest test_zero_redundancy_optimizer.py -q -k test_ddp_zero_overlap
```

To isolate the fix from MIOpen cache state, the trigger was also injected directly by offsetting each rank's inputs by rank * 1e-6, which is what real DDP training looks like. All 16 parametrizations then failed before the change and passed after it, and the 2-rank control passed bitwise either way while 4 and 8 ranks failed at 70.4%, matching the reported rank dependency.

With the change applied the parity is bitwise exact, 0 of 25557032 elements differing across all 16 parametrizations. To confirm that warming the reference removed a false positive rather than blunting the assertion, a deliberate bug was injected into the ZeRO hook (_perform_local_step skipping bucket 2, so one shard is never updated); the fixed test still fails it at 160 of 161 tensors and max_abs_diff=71.6.

The change was authored with the assistance of an AI coding agent; all code and test results were reviewed by me.

### CODE REVIEW

**Review by jeffdaily — APPROVED** [review]
Root cause matches the code: both DDP instances take the `[sys.maxsize]` initial-bucket path, so before this change the reference's first compared iteration was a single all-reduce while the overlap model was already on rebuilt buckets. The warmup adds no optimizer steps and `zero_grad()` precedes every compared iteration, so the reference trajectory is unchanged apart from bucket state. All 16 parametrizations ran on the 4-GPU mi300 distributed shards.
